### ISSUE TITLE
Redirecting factories can have non-required, non-nullable named parameters

### ISSUE THREAD

**Issue by fishythefish** [dart-lang/sdk#52524]

**Comment by eernstg** [comment]
@rrousselGit, those examples are actually a bit different: With the factory constructor we can invoke `Foo()` which will pass null as the value of the formal parameter `x`, but `Foo(x: null)` is a compile-time error. I don't see any way to call your declarations such that the omission of an optional parameter will cause a parameter to be passed by default with a value which is a violation of the declared parameter type.

```dart
abstract class Foo {
  void method({int a});
}

class Bar implements Foo {
  void method({int? a}) {}
}

class Baz implements Foo {
  void method({int a = 42}) {}
}

void main() {
  Foo foo = Bar();
  foo.method(); // Will pass null, which is OK for a parameter of type `int?`.
  foo = Baz();
  foo.method(); // Will pass 42, which is OK for a parameter of type `int`.
}
```

The declaration of `void method({int a})` just says that there is an optional named parameter and we can pass an `int`. It doesn't say how the actual implementation will provide a default value if it is omitted, but each of them will do that in a type correct way.

The abstract method can never be invoked, it is a given that there will be some overriding declaration that provides an implementation. Each of those just need to be a correct override. It does not make sense to say that `a` in `Foo.method` has a default value: Different invocations from the same call site could yield different values.

However, the redirecting factory constructor is more like a forwarding function: It will accept the actual arguments according to the declared parameters, and the behavior of the (omitted) body is to call the specified redirectee constructor with a specific list of actual arguments. This means that the redirecting factory needs to pass mandatory as well as optional arguments to the redirectee, and the design choice was to enforce that they use the same default values.

However, if the default values for the redirectee won't work for the redirecting factory (as in this example) then we just make it an error to have the redirection in the first place. If you want to use a different default value, or change the actual arguments passed to the other constructor then you just need to use a regular factory constructor, not a redirecting one.

It is true that this choice ("can't change the default") has also been made because this makes it possible to implement the compilation of redirecting factory constructor invocations by inlining at the call site, but that should remain an implementation detail, not something which is assumed in the specification.

**Comment by eernstg** [comment]
@chloestefantsova just took a look. Interestingly, the following very similar example _does_ give rise to a compile-time error:

```dart
abstract class Foo {
  int? get x;

  Foo._();

  factory Foo({int x}) = FooImpl;
}

class FooImpl extends Foo {
  @override
  final num x;

  FooImpl({this.x = 2.5}) : super._();
}
```

The situation is the same: The redirecting factory implicitly gets a default value of `2.5`, but the declared parameter type is `int`. But in this case we _do_ get a compile-time error. So the cases where the error isn't detected are somehow null specific.

@dart-lang/language-team, do you see a way to explain this behavior, do you want to make it a spec change? If you agree with me that there is no reasonable way to specify the current behavior then we would proceed and make this a bug fix implementation request. As mentioned, we may need to consider this as a breaking change.

**Comment by rrousselGit** [comment]
@eernstg That is unexpected to me. I would expect redirecting factories to behave as constructor tear-offs and would expect to be able to upcast parameters if need be.


Now that we have constructor tear-offs, I would expect:

```dart
class Foo {
  factory Foo.named() = Bar;
}
```

To be pretty much equivalent to:

```dart
class Foo {
  // downcast the return value
  static const Foo Function() named = Bar.new;
}
```

Where we could upcast parameters or do other things. But with the added benefits of being able to use `const`, make the constructor "unnamed", and disable the default implicit constructor.


To begin with, redirecting constructors are often used as a means to hide implementation details.

If redirecting constructors aren't expected to have differences from the target constructor, I would expect the syntax instead to be:

```dart
class Foo {
  // No specification of the arguments.
  factory Foo = Bar;
}
```

**Comment by eernstg** [comment]
> I would expect redirecting factories to behave as constructor tear-offs

It is not precisely the same kind of thing: A statically resolved reference to a redirecting factory constructor can generally be inline expanded, which means that it is almost exclusively a compile-time phenomenon. Still, when we've declared a parameter type (in particular, when it differs from the corresponding parameter type of the forwardee), it should be respected.

In contrast, a tear-off is a run-time function object. Function objects have a run-time type, and this is a function type. The function type of a redirecting factory constructor will also be the function type of the tear-off. For example:

```dart
class Foo {
  factory Foo.named() = Bar;
}

class Bar implements Foo {}

void main() {
  print(Foo.named.runtimeType); // 'Foo Function()'.
}
```

With the original example, `FooImpl.new` has type `FooImpl Function({int? x})` and `Foo.new` has type `Foo Function({int x})`.

Function types do not specify default values, it's up to the function object itself to handle invocations where some parameters are omitted (but, of course, the function type does specify which parameters are optional).

As you mention, it would be possible to omit the declaration of the formal parameters in a redirection declaration, like this:

```dart
class Foo {
  // No specification of the arguments.
  factory Foo = Bar;
}
```

This would indeed be a possible enhancement (we could allow the formal parameter list declaration to be omitted, as on option). However, I think the main reason why this declaration is required (this was decided before I joined Google) is that it serves as documentation in the factory declaration (otherwise we'd have to look up the redirectee in order to know anything about how it can be invoked).

Also, when the parameter list is always declared (rather than always being omitted), we get the option to declare different parameter types. This means that a redirection can be more strict on the actual arguments than the redirectee, which can be useful.

**Comment by rrousselGit** [comment]
> Also, when the parameter list is always declared (rather than always being omitted), we get the option to declare different parameter types. This means that a redirection can be more strict on the actual arguments than the redirectee, which can be useful.

I agree with that, but it's this very reason which makes me believe that being unable to upcast `int?` as `int` is unexpected.


There are numerous ways in which a redirecting factory can be more strict than its target constructor. So I'm not convinced by the fact that we shouldn't be able to upcast parameters like this.



It's probably a bit off-topic at this point though. If there is a bug, it should indeed be fixed, even if I disagree with the behavior 😛

**Comment by eernstg** [comment]
> being unable to upcast `int?` as `int` is unexpected.

Right, but it is also untrue:

```dart
abstract class Foo {
  int? get x;

  Foo._();

  factory Foo({int x}) = FooImpl;
}

class FooImpl extends Foo {
  @override
  final int? x;

  FooImpl({this.x = 0}) : super._();
}
```

The error does not occur because it is wrong to have parameter type `int` in `Foo` and parameter type `int?` in `FooImpl`, it occurs because the default value taken from the `FooImpl` constructor cannot be used as the default value in the `Foo` constructor.

**Comment by lrhn** [comment]
> The redirecting factory implicitly gets a default value of 2.5,

I don't believe that's true. At least it's not what is specified, so I hope it's not what's implemented.

> ```latex
> It is a compile-time error if $k$ explicitly specifies
> a default value for an optional parameter.
> 
> \rationale{%
> Default values specified in $k$ would be ignored,
> since it is the \emph{actual} parameters that are passed to $k'$.
> Hence, default values are disallowed.%
> }
> ```

This claims that even if the redirecting factory did get an implicit default value, it would be *ignored*, so it shouldn't get one.

There are no coercions or changes to the argument list between the invocation of the redirecting factory constructor and the invocation of its target, the argument list is forwarded unmodified. That includes not inserting default values for missing arguments until it reaches the target constructor. (We specifically require the parameters of the redirecting constructor to be subtypes of the target constructor parameters, not just assignable, because they must be directly applicable without any fiddling.)


So:
```dart
abstract class Foo {
  int? get x;

  Foo._();

  factory Foo({int x}) = FooImpl;
}

class FooImpl extends Foo {
  @override
  final int? x;

  FooImpl({this.x}) : super._();
}
```
is, as I read it, correct and valid code. Any compile-time errors are bugs in the front-ends.

**Comment by rrousselGit** [comment]
@eernstg 
This is more of a wording issue than a misunderstanding. I've talked about upcasts here because this scenario only happens on upcasts afaik.

In the end, my issue with this is, we can do what appears to be the same thing with methods/tear-offs/...    
Sure, I get that you're arguing that they are different. But I personally don't see it. I see redirecting constructors as simple aliases, akin to a typedef for constructors.

The idea that they are possibly actual functions instead of a simple abstraction is obscure to me.
If that was the case, I wouldn't expect redirected factories to infer default values, but instead require folks to duplicate them:


```dart
abstract class Foo {
  // Users may change the default value if they wish.
  // We may have a lint warning if default values are different.
  factory Foo({int x = 0}) = FooImpl;
}

class FooImpl extends Foo {
  FooImpl({int x = 0});
}
``` 

Then, the discussed scenario would become:

```dart
abstract class Foo {
  // Different default value from FooImpl due to different prototype
  factory Foo({int x = 0}) = FooImpl;
}

class FooImpl extends Foo {
  FooImpl({int? x});
}
```


As it stands, the current behavior seems conflicting to me.

**Comment by eernstg** [comment]
@lrhn wrote:
> The design choice was to forward the _entire argument list_, unchecked and unchanged, to the target constructor,

Granted, the specification is actually somewhat inconsistent on this point.

The error about the type-incorrect default value is specified [here](https://github.com/dart-lang/language/blob/cc2c933b1ae5a2c837f89ca4490040ff158a676f/specification/dartLangSpec.tex#L4282), so we should have the error according to the specification. That is, it is _not_

> correct and valid code.

The specification of the dynamic semantics is basically a description of an inlined function invocation. I think it's perfectly fine that the specification _allows_ inlining of invocations, but it is too much of an implementation detail to _require_ that redirecting factory constructor invocations must be inlined.

It is also a questionable choice because it leaves us with a very vague idea about the desired behavior of a tear-off. You could say that the tear-off should be a function that calls the forwardee constructor with exactly the same actual argument list as it received. However, this is not implementable in Dart.

(For good reasons, we don't have the magic `?` operator which we once had, which was able to determine whether or not a given optional argument was passed or not. So we can't reliably make the distinction between having received a default value because it was passed, or having received it because nothing was passed).

Then you could say that the tear-off should also be inlined (that is, if we tear off `Foo` then we should get the same result as a tearoff of `FooImpl` would yield).

We don't do that, and it seems very reasonable to me that we don't do that. We actually get a tear-off whose type is `Foo Function({int x})`, not `FooImpl Function({int? x})`.

I find it hard to explain what the tear-off actually _is_, other than this, and that's a compile-time error:

```dart
Foo tearoffOfFoo({int x = null}) => FooImpl(x: x);
```

**Comment by lrhn** [comment]
> The error about the type-incorrect default value is specified [here](https://github.com/dart-lang/language/blob/cc2c933b1ae5a2c837f89ca4490040ff158a676f/specification/dartLangSpec.tex#L4282), 

That is pretty clear. There is no argument for the line, and I don't remember why it was added.

Let's check. Nope, no justification visible in the commit comment (and I reviewed it).
https://dart-review.googlesource.com/c/sdk/+/71140/5/docs/language/dartLangSpec.tex#2100

I don't see any discussions from around there which can explain why the restriction was added,
other than perhaps a wish to make forwarding factory constructors implementable by desugaring it into a non-redirecting factory constructor `=> TargetConstructor(args);`. 
But that's not how it's actually implemented (I think, haven't checked the dev-compiler, but IIRC, redirecting factory constructors are desugared away in the front-end, forwarding the call directly to the target constructor after satisfying the static type checks of the redirecting constructor). The VM relied on that behavior until very recently in it's implementation of the `List` constructor.

The problem with starting to enforce this restriction is:
* Not what's implemented. 
* Not as useful (more restrictive, with no actual benefit), and
* Not how I thought things worked (very important! :grin: )

The line was added in 2018, we have disallowed default values since, at least [2013](https://github.com/dart-lang/sdk/issues/14284). 

So I'd remove that line again, it's not necessary, and we don't use it today.

**Comment by eernstg** [comment]
> I don't remember why it was added.

Same here, but I would be quite surprised if the motivation had anything to do with a possible implementation strategy.

I still think it's a surprising anomaly, based on what I said earlier about the original example:

> Presumably, `Foo` has the parameter type `int` because it is intended to be enforced that the value of that parameter is an `int` for all invocations of that constructor. If it is intended to be allowed to be null then the parameter type in `Foo` might as well have been `int?`.

It is also an anomaly that `Foo(x: null)` is a compile time error, and `Foo(x: null as dynamic)` is a run-time error, but `Foo()` will pass null. It's a mess. It can only be explained if we insist on talking about inlined invocations, and that's an implementation detail. Finally, the tear-off doesn't work the same as the constructor:

```dart
abstract class Foo {
  int? get x;
  Foo._(Object? x);
  factory Foo({int x}) = FooImpl;
}

class FooImpl extends Foo {
  @override
  final int? x;
  FooImpl({this.x}) : super._(x);
}

void main() {
  Function f = Foo.new;
  f(); // Throws, "null is not an `int`".
}
```

**Comment by lrhn** [comment]
About the anomaly, try comparing it to the following:

```dart
class D {
  void foo([int? x = 42]) {
    print("x = $x");
  }
}
class E extends D {
  void foo([covariant int x]);
}
void main() {
  // E().foo(null); // compile-time error 
  // E().foo(null as dynamic); // run-time error
  E().foo(); // works
  (E() as D).foo(); // works too, cannot do that for constructor
  (E() as D).foo(null); // works, ditto
}
```

Here we have a statically defined type signature, in `E`, which is implemented by a method with a more permissive parameter list.
We get the same static and runtime errors as for the constructor, but we can still omit the argument and get the default value of the actual implementation.

**Comment by eernstg** [comment]
> We get the same static and runtime errors as for the constructor, but we can still omit the argument and get the default value of the actual implementation

Not quite:

When an instance member implementation has an optional parameter (like `D.foo.x`) and it is invoked based on a static receiver type whose statically known member of that name is an abstract declaration, the abstract declaration can omit or include the default value declaration, but the actually running declaration determines the default value.

We're consistently maintaining that an abstract member declaration does not correspond to an actual function, and the same call site could be called several times and get different actual implementations each time, possibly with different default values. Those implementations could widen the parameter type and thus use a default value which isn't type correct for the statically known signature, but that signature is in any case just a conservative approximation of the signature of the member implementation that we're actually calling.

With the redirecting factory constructor and the redirectee constructor, there is no dynamic dispatch. The situation is more like this one:

```dart
FooImpl redirectee({int? x}) => ...;
Foo redirecting({int x = /*generated*/}) => redirectee(x: x);
```

The declaration of `redirecting` uses a default value which is implicitly induced by the compiler/analyzer. It isn't type correct to use null, but that is the default value that we will have if it is copied from `redirectee`.

The immediate counter-argument would be that invocations are inlined (they're even specified like that), and there is no actual `redirecting` function (that is, there is no actual `Foo` constructor). So we don't care about that default value, because we simply call `FooImpl(with, any, given, arguments)`. When no arguments are passed, as in `Foo()`, it just means that we invoke `FooImpl()`, and that expression already has a well-defined meaning (using its own default value).

This perspective will work for invocations of the constructors, but it doesn't work as well for tear-offs.

In the original example, we could tear off constructors using `Foo.new` and `FooImpl.new`. The resulting function objects have different return types (`Foo.new` has type `Foo Function({int x})` and `FooImpl.new` has type `FooImpl Function({int? x})`), which implies that the torn off function isn't simply "inlined" (that is, `Foo.new` and `FooImpl.new` do not simply return the same function).

I think this _is_ the appropriate behavior for the tear-off operation. However, if we wish to explain what is going on then we'd need to talk about the two constructors as entities that exist, which gives rise to the perspective that they are similar to the functions `redirectee` and `redirecting` above, which again gives rise to the question: Which default value does `Foo.new` actually have?

I'm simply saying that it should have the default value which is also the default value of the redirectee (with `Foo` and `FooImpl` this value is null), and it should be a compile-time error if that default value isn't type correct (in the example, null must have type `int`, which it doesn't, so that's an error).

**Comment by lrhn** [comment]
I actually do think of the redirecting factory constructor as a not-quite-real static type applied to the actual constructor it forwards to, one which isn't there in practice.

I can see the problem with constructor tear-offs, though. We usually say that a constructor tear-off can be seen as equivalent to tearing off a "corresponding static function" (suitably instantiated for generic classes), but we cannot write a Dart function which behaves like the tear-off of a forwarding factory constructor which forwards "no value" without having the same default value for its argument. So if the forwarding function's parameter has a type which precludes that default value, it's just not writable in Dart.

That doesn't mean it can't work that way, just that the behavior cannot be desugared. (Which is a problem when the specification of constructor tear-offs uses desugaring to define the behavior.)

And it doesn't work everywhere:
```dart
class C {
  final num? x;
  C([num? x = 42.5]) : x = x;
  factory C.r([int x]) = C;
}

void main() {
  var f = C.r;
  print(C.r.runtimeType);  // ([int]) => C
  print(C.r().x);          // 42.5
  print(f().x);            // ?
}
```
This first prints `([int]) => C` and `42.5`  in both dart2js and the VM.
Then dart2js fails with "Uncaught Error: TypeError: 42.5: type 'double' is not a subtype of type 'int'", while the VM prints `42.5` again.
Our implementations do not agree on how the tear-off is implemented.
We should probably fix that.

(It also shows that it's *possible* to implement it as forwarding, even if it cannot be expressed in Dart code, since the VM does so.)

**Comment by eernstg** [comment]
> It also shows that it's _possible_ to implement it as forwarding

Sure, we can use a magic sentinel value whose run-time type is considered to be `Never` (which will only occur in very well-understood situations, because that value cannot be denoted by user code; it's basically a private, extra null):

```dart
void f({int? x}) {...}
void g({int x = magicSentinel}) => f(x: x == magicSentinel ? null : x);
```

I just don't think it is _helpful_ to anyone to work so hard just so `g` can work as if it had had a default value which isn't type correct.

**Comment by rrousselGit** [comment]
In my opinion the value is in not having clunky edge-cases when the syntax is used in multiple places.

Supporting this officially, when considering that methods can do something similar, would be a bit more consistent
